outscale/resource_security_group.go

```diff
@@ -0,0 +1,554 @@
+		if err != nil {
+			return utils.CheckThrottling(httpResp, err)
+		}
+		createResp = rp
+		return nil
+	})
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to create Security Group.",
+			err.Error(),
+		)
+		return
+	}
+	sg := createResp.GetSecurityGroup()
+	data.RequestId = types.StringValue(createResp.ResponseContext.GetRequestId())
+	data.Id = types.StringValue(sg.GetSecurityGroupId())
+	data.SecurityGroupId = types.StringValue(sg.GetSecurityGroupId())
+
+	if data.RemoveDefaultOutboundRule.ValueBool() {
+		ipRange := "0.0.0.0/0"
+		ipProtocol := "-1"
+		emptySGReq := oscgo.DeleteSecurityGroupRuleRequest{
+			Flow:            "Outbound",
+			SecurityGroupId: sg.GetSecurityGroupId(),
+			IpRange:         &ipRange,
+			IpProtocol:      &ipProtocol,
+		}
+
+		deleteTimeout, diag := data.Timeouts.Delete(ctx, utils.DeleteDefaultTimeout)
+		if utils.CheckDiags(resp, diag) {
+			return
+		}
+		ctx, cancel := context.WithTimeout(ctx, deleteTimeout)
+		defer cancel()
+
+		err := retry.RetryContext(ctx, deleteTimeout, func() *retry.RetryError {
+			_, httpResp, err := r.Client.SecurityGroupRuleApi.DeleteSecurityGroupRule(ctx).DeleteSecurityGroupRuleRequest(emptySGReq).Execute()
+			if err != nil {
+				return utils.CheckThrottling(httpResp, err)
+			}
+			return nil
+		})
+		if err != nil {
+			resp.Diagnostics.AddError(
+				"Unable to empty the Security Group rules.",
+				err.Error(),
+			)
+			return
+		}
+	}
+
+	if len(data.Tags) > 0 {
+		err = createFrameworkTags(ctx, r.Client, tagsToOSCResourceTag(data.Tags), sg.GetSecurityGroupId())
+		if err != nil {
+			resp.Diagnostics.AddError(
+				"Unable to add Tags on the Security Group resource.",
+				err.Error(),
+			)
+			return
+		}
+	}
+
+	stateData, err := setSecurityGroupState(ctx, r, data)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to set Security Group state.",
+			err.Error(),
+		)
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &stateData)...)
+}
+
+func (r *resourceSecurityGroup) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var data SecurityGroupModel
+	diag := req.State.Get(ctx, &data)
+	if utils.CheckDiags(resp, diag) {
+		return
+	}
+
+	data, err := setSecurityGroupState(ctx, r, data)
+	if err != nil {
+		if err.Error() == "Empty" {
+			resp.State.RemoveResource(ctx)
+			return
+		}
+		resp.Diagnostics.AddError(
+			"Unable to set Security Group API response values.",
+			err.Error(),
+		)
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *resourceSecurityGroup) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var planData, stateData SecurityGroupModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &planData)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &stateData)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if !reflect.DeepEqual(planData.Tags, stateData.Tags) {
```

Comment:
Are you sure you get always the same order ? If you have a different order, your test fails.

Note: I guess tags are comparable, you could use `slices.Equal` if that's the case.

You could also check `lo.ElementsMatch` in samber/lo to have a func that does not expect the right order.

Or always compute the diff, and only update when toRemove or toCreate is not empty.